### DISCUSSION TITLE
Rerenders with useFieldArray and form.reset()

### DISCUSSION THREAD

**Discussion by AdrianMrn** [react-hook-form/react-hook-form#11770] [Q&A]
After I submit my form, I call `form.reset(valuesFromApi)` to set the new values as my defaultValues, and be able to set my dirty state, this works as expected. However, this is causing all my components that use `useFieldArray` to remount completely, which causes the internal keys of the items to be recalculated, and all the children to unmount and remount as well, which causes lost focus in input fields etc.

I have an autosave which runs every couple of seconds, which is why this loss of focus is such an issue for me. I'm currently working around it by disabling autosave when the user has focus in an input or textarea, but it doesn't feel correct having this, and quite fragile. Eg file inputs also suffer from this, and won't set the value properly if the item has remounted in the meantime.

I've used `shouldUnregister: false`, which doesn't fix my issue.

**Comment by mir0ir** [comment]
Also having this issue!

**Reply by frikkfelix** [reply]
+1

**Comment by KarlinskyS** [comment]
I faced a similar issue, but from a slightly different perspective. I was using useFieldArray, and part of my rendered values (specifically const { fields } = useFieldArray({ control, name: 'someSpecificForm' })) were tied to inputs managed separately via useState. When I called reset(), the input that **wasn't tied to react-hook-form** was also unexpectedly reset to its initial value.

After some investigation, I realized that reset() causes useFieldArray's **mySuperField** to unmount and remount components. This unmounting affects all child components, including those managed by useState, which explains why my input's state reverted to its initial value (null in my case) after the remount.

I hope this helps others facing similar issues!

small code snippet:

```
  const MyForm = () => {
  const { control, reset } = useForm({
    defaultValues: {
      items: [{ name: "Item 1" }, { name: "Item 2" }]
    }
  });

  // useFieldArray to dynamically render items
  const { fields } = useFieldArray({
    control,
    name: "items"
  });

  // A separate state-controlled input
  const [externalInput, setExternalInput] = useState("External State");

  const handleReset = () => {
    reset(); // This triggers unmount/remount of useFieldArray components
    // externalInput gets reset because of the remount cycle
  };

  return (
    <div>
      <button type="button" onClick={handleReset}>
        Reset Form
      </button>

      {/* useFieldArray renders dynamically */}
      {fields.map((field, index) => (
        <div key={field.id}>
          <input
            name={`items[${index}].name`}
            defaultValue={field.name}
          />
        </div>
      ))}

      {/* An input controlled by useState */}
      <input
        value={externalInput}
        onChange={(e) => setExternalInput(e.target.value)}
        placeholder="External Input"
      />
    </div>
  );
};
```

**Comment by tevinch** [comment]
For the save-acknowledgement part of this flow, current React Hook Form has a public [`resetDefaultValues`](https://react-hook-form.com/docs/useform/resetdefaultvalues) method. It updates the baseline used for `isDirty`/`dirtyFields` while keeping the current values and field-array rows in place.

The essential step inside a validated save callback is:

```tsx
const submittedSnapshot = structuredClone(values);
const acknowledgedValues = await saveSnapshot(submittedSnapshot);
resetDefaultValues(acknowledgedValues);
```

If the user types again while that request is pending, the later text stays visible and remains dirty against the saved snapshot. Pass the values that were actually acknowledged; reading current form values at response time could mark newer, unsaved edits as saved.

I made a small [MIT component and runnable example](https://github.com/tevinch/data-shape-kit/tree/main/examples/saved-field-array), tested with React Hook Form 7.87.0, covering retained row state, focus, in-flight edits and failure/retry. The example uses an explicit slow local save to make the timing visible; the same acknowledgement step fits an existing autosave callback. It updates the baseline only, so applying server-normalized values or merging background changes remains a separate application decision.
